File: backend/services/roleService.js
// backend/services/roleService.js
import db from "../config/db.js";
import AppError from "../utils/AppError.js";
import * as roleRepo from "../repositories/roleRepository.js";
import { createLog } from "../repositories/systemLogRepository.js";
import { emitSharedTaskSignal } from "./firebaseSignalService.js";
import cache from "../config/cache.js";
import Logger from "../utils/logger.js";

/**
 * @returns {Promise<any>}
 */
export const getRolesService = async () => {
  const cacheKey = "MASTER_ROLES";
  if (cache.has(cacheKey)) return cache.get(cacheKey);

  const data = await roleRepo.getRoles(db);
  cache.set(cacheKey, data);
  return data;
};

/**
 * @returns {Promise<any>}
 */
export const getPermissionsService = async () => {
  const cacheKey = "MASTER_PERMISSIONS";
  if (cache.has(cacheKey)) return cache.get(cacheKey);

  const data = await roleRepo.getPermissions(db);
  cache.set(cacheKey, data);
  return data;
};

/**
 * @param {number|string} roleId
 * @returns {Promise<any>}
 */
export const getRolePermissionsService = async (roleId) => {
  return await roleRepo.getRolePermissions(db, roleId);
};

/**
 * @param {number|string} roleId
 * @param {number|string} permissionIds
 * @param {number|string} userId
 * @param {any} ip
 * @param {any} userAgent
 * @returns {Promise<any>}
 */
export const updateRolePermissionsService = async (roleId, permissionIds, userId, ip, userAgent) => {
  let connection;
  try {
    connection = await db.getConnection();
    await connection.beginTransaction();

    await roleRepo.deleteRolePermissions(connection, roleId);
    await roleRepo.insertRolePermissions(connection, roleId, permissionIds);

    await createLog(connection, {
      userId: userId,
      action: "UPDATE",
      targetType: "ROLE",
      targetId: String(roleId),
      changes: { note: "Updated Role Permissions", permissionCount: permissionIds.length },
      ip,
      userAgent,
    });

    await connection.commit();
    cache.del("MASTER_ROLES");
    emitSharedTaskSignal('MASTER_DATA', 'REFRESH_ROLES').catch(e => Logger.error("Signal Error", e, "ROLE_SERVICE"));
  } catch (error) {
    if (connection) await connection.rollback();
    if (error.code === "ER_NO_REFERENCED_ROW_2") {
      throw new AppError("Satu atau lebih ID izin atau peran tidak valid.", 400);
    }
    throw error;
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {any} name
 * @param {any} description
 * @param {number|string} userId
 * @param {any} ip
 * @param {any} userAgent
 * @returns {Promise<any>}
 */
export const createRoleService = async (name, description, userId, ip, userAgent) => {
  try {
    const roleId = await roleRepo.createRole(db, name, description);

    await createLog(db, {
      userId,
      action: "CREATE",
      targetType: "ROLE",
      targetId: String(roleId),
      changes: { name, description },
      ip,
      userAgent,
    });

    cache.del("MASTER_ROLES");
    emitSharedTaskSignal('MASTER_DATA', 'REFRESH_ROLES').catch(e => Logger.error("Signal Error", e, "ROLE_SERVICE"));
    return roleId;
  } catch (error) {
    if (error.code === "ER_DUP_ENTRY") {
      throw new AppError("Nama peran sudah ada.", 409);
    }
    throw error;
  }
};

/**
 * @param {number|string} roleId
 * @param {any} name
 * @param {any} description
 * @param {number|string} userId
 * @param {any} ip
 * @param {any} userAgent
 * @returns {Promise<any>}
 */
export const updateRoleService = async (roleId, name, description, userId, ip, userAgent) => {
  try {
    const isUpdated = await roleRepo.updateRole(db, roleId, name, description);
    if (!isUpdated) throw new AppError("Peran tidak ditemukan.", 404);

    await createLog(db, {
      userId,
      action: "UPDATE",
      targetType: "ROLE",
      targetId: String(roleId),
      changes: { name, description },
      ip,
      userAgent,
    });
    
    cache.del("MASTER_ROLES");
    emitSharedTaskSignal('MASTER_DATA', 'REFRESH_ROLES').catch(e => Logger.error("Signal Error", e, "ROLE_SERVICE"));
  } catch (error) {
    if (error.code === "ER_DUP_ENTRY") {
      throw new AppError("Nama peran sudah digunakan.", 409);
    }
    throw error;
  }
};

/**
 * @param {number|string} roleId
 * @param {number|string} userId
 * @param {any} ip
 * @param {any} userAgent
 * @returns {Promise<any>}
 */
export const deleteRoleService = async (roleId, userId, ip, userAgent) => {
  try {
    const isDeleted = await roleRepo.deleteRole(db, roleId);
    if (!isDeleted) throw new AppError("Peran tidak ditemukan.", 404);

    await createLog(db, {
      userId,
      action: "DELETE",
      targetType: "ROLE",
      targetId: String(roleId),
      changes: { note: "Deleted Role" },
      ip,
      userAgent,
    });
    
    cache.del("MASTER_ROLES");
    emitSharedTaskSignal('MASTER_DATA', 'REFRESH_ROLES').catch(e => Logger.error("Signal Error", e, "ROLE_SERVICE"));
  } catch (error) {
    if (error.code === "ER_ROW_IS_REFERENCED_2") {
      throw new AppError("Gagal menghapus: Peran ini masih digunakan oleh satu atau lebih pengguna.", 400);
    }
    throw error;
  }
};

File: backend/internal/service/role_service.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"strconv"
	"strings"

	"github.com/dps-wmhris/backend/internal/database"
	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/model"
	"github.com/dps-wmhris/backend/internal/repository"
	"github.com/jmoiron/sqlx"
)

type RoleService interface {
	GetRoles(ctx context.Context) ([]dto.RoleResponse, error)
	GetPermissions(ctx context.Context) ([]dto.PermissionResponse, error)
	GetRolePermissions(ctx context.Context, roleID int) ([]int, error)
	UpdateRolePermissions(ctx context.Context, roleID int, req dto.AssignPermissionsRequest, userID int, ip, userAgent string) error
	CreateRole(ctx context.Context, req dto.CreateRoleRequest, userID int, ip, userAgent string) (int, error)
	UpdateRole(ctx context.Context, roleID int, req dto.CreateRoleRequest, userID int, ip, userAgent string) error
	DeleteRole(ctx context.Context, roleID int, userID int, ip, userAgent string) error
}

type roleServiceImpl struct {
	db            *sqlx.DB
	roleRepo      repository.RoleRepository
	systemLogRepo repository.SystemLogRepository
}

func NewRoleService(db *sqlx.DB, roleRepo repository.RoleRepository, systemLogRepo repository.SystemLogRepository) RoleService {
	return &roleServiceImpl{
		db:            db,
		roleRepo:      roleRepo,
		systemLogRepo: systemLogRepo,
	}
}

func (s *roleServiceImpl) GetRoles(ctx context.Context) ([]dto.RoleResponse, error) {
	return s.roleRepo.GetRoles(ctx)
}

func (s *roleServiceImpl) GetPermissions(ctx context.Context) ([]dto.PermissionResponse, error) {
	return s.roleRepo.GetPermissions(ctx)
}

func (s *roleServiceImpl) GetRolePermissions(ctx context.Context, roleID int) ([]int, error) {
	return s.roleRepo.GetRolePermissions(ctx, roleID)
}

func (s *roleServiceImpl) UpdateRolePermissions(ctx context.Context, roleID int, req dto.AssignPermissionsRequest, userID int, ip, userAgent string) error {
	err := database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		err := s.roleRepo.DeleteRolePermissions(ctx, tx, roleID)
		if err != nil {
			return err
		}

		err = s.roleRepo.InsertRolePermissions(ctx, tx, roleID, req.PermissionIDs)
		if err != nil {
			return err
		}

		return nil
	})

	if err != nil {
		if strings.Contains(err.Error(), "1452") || strings.Contains(err.Error(), "foreign key constraint fails") {
			return errors.New("Role atau Permission ID tidak valid.")
		}
		return err
	}

	// Logging
	changes := map[string]interface{}{
		"note":            "Updated Role Permissions",
		"permissionCount": len(req.PermissionIDs),
	}
	changesBytes, _ := json.Marshal(changes)
	changesStr := string(changesBytes)
	
	var ipPtr, uaPtr *string
	if ip != "" { ipPtr = &ip }
	if userAgent != "" { uaPtr = &userAgent }

	s.systemLogRepo.Create(ctx, s.db, &model.SystemLog{
		UserID:     userID,
		Action:     "UPDATE",
		TargetType: "ROLE",
		TargetID:   strconv.Itoa(roleID),
		Changes:    &changesStr,
		IP:         ipPtr,
		UserAgent:  uaPtr,
	})

	return nil
}

func (s *roleServiceImpl) CreateRole(ctx context.Context, req dto.CreateRoleRequest, userID int, ip, userAgent string) (int, error) {
	roleID, err := s.roleRepo.CreateRole(ctx, req.Name, req.Description)
	if err != nil {
		if strings.Contains(err.Error(), "1062") || strings.Contains(err.Error(), "Duplicate entry") {
			return 0, errors.New("Nama peran sudah digunakan.")
		}
		return 0, err
	}

	changes := map[string]interface{}{
		"name":        req.Name,
		"description": req.Description,
	}
	changesBytes, _ := json.Marshal(changes)
	changesStr := string(changesBytes)
	
	var ipPtr, uaPtr *string
	if ip != "" { ipPtr = &ip }
	if userAgent != "" { uaPtr = &userAgent }

	s.systemLogRepo.Create(ctx, s.db, &model.SystemLog{
		UserID:     userID,
		Action:     "CREATE",
		TargetType: "ROLE",
		TargetID:   strconv.Itoa(roleID),
		Changes:    &changesStr,
		IP:         ipPtr,
		UserAgent:  uaPtr,
	})

	return roleID, nil
}

func (s *roleServiceImpl) UpdateRole(ctx context.Context, roleID int, req dto.CreateRoleRequest, userID int, ip, userAgent string) error {
	isUpdated, err := s.roleRepo.UpdateRole(ctx, roleID, req.Name, req.Description)
	if err != nil {
		if strings.Contains(err.Error(), "1062") || strings.Contains(err.Error(), "Duplicate entry") {
			return errors.New("Nama peran sudah digunakan.")
		}
		return err
	}
	if !isUpdated {
		return errors.New("peran tidak ditemukan")
	}

	changes := map[string]interface{}{
		"name":        req.Name,
		"description": req.Description,
	}
	changesBytes, _ := json.Marshal(changes)
	changesStr := string(changesBytes)
	
	var ipPtr, uaPtr *string
	if ip != "" { ipPtr = &ip }
	if userAgent != "" { uaPtr = &userAgent }

	s.systemLogRepo.Create(ctx, s.db, &model.SystemLog{
		UserID:     userID,
		Action:     "UPDATE",
		TargetType: "ROLE",
		TargetID:   strconv.Itoa(roleID),
		Changes:    &changesStr,
		IP:         ipPtr,
		UserAgent:  uaPtr,
	})

	return nil
}

func (s *roleServiceImpl) DeleteRole(ctx context.Context, roleID int, userID int, ip, userAgent string) error {
	isDeleted, err := s.roleRepo.DeleteRole(ctx, roleID)
	if err != nil {
		return err // ER_ROW_IS_REFERENCED_2 handling needed later
	}
	if !isDeleted {
		return errors.New("peran tidak ditemukan")
	}

	changes := map[string]interface{}{
		"note": "Deleted Role",
	}
	changesBytes, _ := json.Marshal(changes)
	changesStr := string(changesBytes)
	
	var ipPtr, uaPtr *string
	if ip != "" { ipPtr = &ip }
	if userAgent != "" { uaPtr = &userAgent }

	s.systemLogRepo.Create(ctx, s.db, &model.SystemLog{
		UserID:     userID,
		Action:     "DELETE",
		TargetType: "ROLE",
		TargetID:   strconv.Itoa(roleID),
		Changes:    &changesStr,
		IP:         ipPtr,
		UserAgent:  uaPtr,
	})

	return nil
}
